deps/rabbit/src/rabbit_khepri.erl

```diff
@@ -1414,35 +1426,78 @@ register_rabbit_bindings_projection() ->
+                  pattern = #exchange{type = '$1', _ = '_'},
+                  conditions = [{'andalso',
+                                 {'=/=', '$1', headers},
+                                 {'=/=', '$1', topic},
+                                 {'=/=', '$1', fanout},
+                                 {'=/=', '$1', 'x-jms-topic'},
+                                 {'=/=', '$1', 'x-random'}
+                                }]},
     PathPattern = rabbit_db_binding:khepri_route_path(
                     _VHost = ?KHEPRI_WILDCARD_STAR,
-                    _Exchange = DirectOrFanout,
+                    Exchange,
+                    _Kind = ?KHEPRI_WILDCARD_STAR,
+                    _DstName = ?KHEPRI_WILDCARD_STAR,
+                    _RoutingKey = ?KHEPRI_WILDCARD_STAR),
+    khepri:register_projection(?STORE_ID, PathPattern, Projection).
+
+register_rabbit_route_by_source_projection() ->
+    MapFun = fun(_Path, #binding{source = Source,
+                                 key = Key,
+                                 destination = Destination,
+                                 args = Args}) ->
+                     #route_by_source{source = Source,
+                                      key = Key,
+                                      destination = Destination,
+                                      args = Args}
+             end,
+    ProjectionFun = projection_fun_for_sets(MapFun),
+    Options = #{type => bag,
+                keypos => #route_by_source.source,
+                read_concurrency => true},
+    Projection = khepri_projection:new(rabbit_khepri_route_by_source,
+                                       ProjectionFun,
+                                       Options),
+    %% For some exchange types we know that they won't use this projection.
+    %% So we exclude such bindings for two reasons:
+    %% 1. Lower overall ETS memory usage
+    %% 2. "Avoid inserting an extensive amount of objects with the same key.
+    %%    It will hurt insert and lookup performance as well as real time characteristics
+    %%    of the runtime environment (hash bucket linear search do not yield)."
+    %%    Example: same source direct exchange with 100k different binding keys.
+    %% In future, rather than exchange types exclusion as done here, a nicer approach
+    %% would be that each exchange requiring routing lookup by only source exchange
+    %% advertises this access pattern, e.g. as a boolean flag in the #exchange.options field.
+    Exchange = #if_data_matches{
+                  pattern = #exchange{type = '$1', _ = '_'},
+                  conditions = [{'andalso',
+                                 {'=/=', '$1', headers},
+                                 {'=/=', '$1', topic},
+                                 {'=/=', '$1', direct},
+                                 {'=/=', '$1', 'x-local-random'},
+                                 {'=/=', '$1', 'x-jms-topic'}
+                                }]},
```

Comment:
The approach in this PR is good enough in my opinion.

To implement what I suggested in the code comment, we'd need yet another migration of exchange records, which I would like to avoid for now.

Yet another approach I can think of is adding a new required callback to `rabbit_exchange_type` such that every exchange type implementation has to advertise its routing lookup/access pattern. However, it seems they can't be used in these Khepri trigger conditions. For example, writing the following trigger condition isn't supported:
```erl
rabbit_exchange:route_access_pattern(ExchangeType) =:= source
```

Reply:
Ok, let's keep the current approach then, and leave the comment in the code as a reminder for a future optimisation.